From 3ffbd132300e29a8a94766ea819b18ea5b6c3d20 Mon Sep 17 00:00:00 2001
From: Sagar Cheluvegowda <quic_scheluve@quicinc.com>
Date: Fri, 6 Dec 2024 19:00:58 -0800
Subject: [PATCH] net: stmmac: dwmac-qcom-ethqos: Enable SCMI ETH and deep
 sleep support

On sa8255p platforms, resource such as clocks, regulators etc.
are managed in the firmware. The Linux driver uses power and
performance domains to request resource operations. The domains
are turned on/off using runtime PM framework.

Power domains represents suppliers to put device into various
power state. Two modeled power states d1 and d3 are supported.
There are two power domain suppliers named "core" and "mdio".

D0 <--> D3 :: Turn on/off both suppliers
D0 <--> D1 :: Turn on/off 'mdio' supplier

Performance domain represents the serdes phy with link speed
being used as performance levels.

Add DS/QB support to turn off resources on deep sleep and
turn on quick boot. Exit called on remove & suspend.
Init called on resume.

Change-Id: I916bfa1da2aacd5a540360c77bd73cbde8175336
Signed-off-by: Sarosh Hasan <quic_sarohasa@quicinc.com>
Signed-off-by: Sagar Cheluvegowda <quic_scheluve@quicinc.com>
---
 .../stmicro/stmmac/dwmac-qcom-ethqos.c        | 224 ++++++++++++++++--
 1 file changed, 210 insertions(+), 14 deletions(-)

diff --git a/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c b/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c
index 1ed01f975dfd..61a0d1249231 100644
--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-qcom-ethqos.c
@@ -7,6 +7,8 @@
 #include <linux/platform_device.h>
 #include <linux/phy.h>
 #include <linux/phy/phy.h>
+#include <linux/pm_opp.h>
+#include <linux/pm_domain.h>
 #include <linux/pcs-xpcs-qcom.h>
 
 #include "stmmac.h"
@@ -110,6 +112,12 @@
 
 #define SGMII_10M_RX_CLK_DVDR			0x31
 
+enum domain_t {
+	POWER_CORE = 0,
+	POWER_MDIO = 1,
+	PERF_SERDES = 2,
+};
+
 struct ethqos_emac_por {
 	unsigned int offset;
 	unsigned int value;
@@ -124,6 +132,7 @@ struct ethqos_emac_driver_data {
 	const char *link_clk_name;
 	bool has_integrated_pcs;
 	u32 dma_addr_width;
+	unsigned int ptp_clk_rate;
 	struct dwmac4_addrs dwmac4_addrs;
 	bool needs_sgmii_loopback;
 	bool has_hdma;
@@ -147,6 +156,8 @@ struct qcom_ethqos {
 	bool rgmii_config_loopback_en;
 	bool has_emac_ge_3;
 	bool needs_sgmii_loopback;
+	bool use_domains;
+	struct dev_pm_domain_list *pd_list;
 };
 
 static int rgmii_readl(struct qcom_ethqos *ethqos, unsigned int offset)
@@ -336,6 +347,7 @@ static const struct ethqos_emac_driver_data emac_v4_0_0_data = {
 	.has_integrated_pcs = true,
 	.needs_sgmii_loopback = true,
 	.dma_addr_width = 36,
+	.ptp_clk_rate = 230400000,
 	.dwmac4_addrs = {
 		.dma_chan = 0x00008100,
 		.dma_chan_offset = 0x1000,
@@ -373,6 +385,52 @@ static const struct ethqos_emac_driver_data emac_v6_6_0_data = {
 	},
 };
 
+static int qcom_ethqos_domain_attach(struct qcom_ethqos *ethqos)
+{
+	struct dev_pm_domain_attach_data pd_data = {
+		.pd_flags = PD_FLAG_NO_DEV_LINK,
+		.pd_names = (const char*[]) { "power_core", "power_mdio", "perf_serdes" },
+		.num_pd_names = 3,
+	};
+	struct device *dev = &ethqos->pdev->dev;
+
+	return dev_pm_domain_attach_list(dev, &pd_data, &ethqos->pd_list);
+}
+
+static int qcom_ethqos_domain_on(struct qcom_ethqos *ethqos, enum domain_t dom)
+{
+	struct device *dev = ethqos->pd_list->pd_devs[dom];
+	int ret = 0;
+
+	ret = pm_runtime_resume_and_get(dev);
+	if (ret < 0)
+		dev_err(dev, "poweron(domain=%d) failed.(err=%d)\n", dom, ret);
+
+	return ret;
+}
+
+static void qcom_ethqos_domain_off(struct qcom_ethqos *ethqos, enum domain_t dom)
+{
+	struct device *dev = ethqos->pd_list->pd_devs[dom];
+	int ret = 0;
+
+	ret = pm_runtime_put_sync(dev);
+	if (ret < 0)
+		dev_err(dev, "poweroff(domain=%d) failed.(err=%d)\n", dom, ret);
+}
+
+static int qcom_ethqos_serdes_set_level(struct qcom_ethqos *ethqos)
+{
+	struct device *dev = ethqos->pd_list->pd_devs[PERF_SERDES];
+	int ret = 0;
+
+	ret = dev_pm_opp_set_level(dev, ethqos->speed);
+	if (ret)
+		dev_err(dev, "Failed to set serdes level. (err=%d)\n", ret);
+
+	return ret;
+}
+
 static int ethqos_dll_configure(struct qcom_ethqos *ethqos)
 {
 	struct device *dev = &ethqos->pdev->dev;
@@ -701,8 +759,12 @@ static int ethqos_configure_sgmii(struct qcom_ethqos *ethqos)
 		rgmii_updatel(ethqos, RGMII_CONFIG2_RGMII_CLK_SEL_CFG,
 			      RGMII_CONFIG2_RGMII_CLK_SEL_CFG,
 			      RGMII_IO_MACRO_CONFIG2);
-		if (ethqos->serdes_speed != SPEED_2500)
-			phy_set_speed(ethqos->serdes_phy, SPEED_2500);
+		if (ethqos->serdes_speed != SPEED_2500) {
+			if (ethqos->use_domains)
+				qcom_ethqos_serdes_set_level(ethqos);
+			else
+				phy_set_speed(ethqos->serdes_phy, SPEED_2500);
+		}
 		ethqos->serdes_speed = SPEED_2500;
 		stmmac_pcs_ctrl_ane(priv, priv->ioaddr, 0, 0, 0);
 		break;
@@ -711,15 +773,23 @@ static int ethqos_configure_sgmii(struct qcom_ethqos *ethqos)
 		rgmii_updatel(ethqos, RGMII_CONFIG2_RGMII_CLK_SEL_CFG,
 			      RGMII_CONFIG2_RGMII_CLK_SEL_CFG,
 			      RGMII_IO_MACRO_CONFIG2);
-		if (ethqos->serdes_speed != SPEED_1000)
-			phy_set_speed(ethqos->serdes_phy, SPEED_1000);
+		if (ethqos->serdes_speed != SPEED_1000) {
+			if (ethqos->use_domains)
+				qcom_ethqos_serdes_set_level(ethqos);
+			else
+				phy_set_speed(ethqos->serdes_phy, SPEED_1000);
+		}
 		ethqos->serdes_speed = SPEED_1000;
 		stmmac_pcs_ctrl_ane(priv, priv->ioaddr, 1, 0, 0);
 		break;
 	case SPEED_100:
 		val |= ETHQOS_MAC_CTRL_PORT_SEL | ETHQOS_MAC_CTRL_SPEED_MODE;
-		if (ethqos->serdes_speed != SPEED_1000)
-			phy_set_speed(ethqos->serdes_phy, SPEED_1000);
+		if (ethqos->serdes_speed != SPEED_1000) {
+			if (ethqos->use_domains)
+				qcom_ethqos_serdes_set_level(ethqos);
+			else
+				phy_set_speed(ethqos->serdes_phy, SPEED_1000);
+		}
 		ethqos->serdes_speed = SPEED_1000;
 		stmmac_pcs_ctrl_ane(priv, priv->ioaddr, 1, 0, 0);
 		break;
@@ -730,8 +800,12 @@ static int ethqos_configure_sgmii(struct qcom_ethqos *ethqos)
 			      FIELD_PREP(RGMII_CONFIG_SGMII_CLK_DVDR,
 					 SGMII_10M_RX_CLK_DVDR),
 			      RGMII_IO_MACRO_CONFIG);
-		if (ethqos->serdes_speed != SPEED_1000)
-			phy_set_speed(ethqos->serdes_phy, ethqos->speed);
+		if (ethqos->serdes_speed != SPEED_1000) {
+			if (ethqos->use_domains)
+				qcom_ethqos_serdes_set_level(ethqos);
+			else
+				phy_set_speed(ethqos->serdes_phy, SPEED_1000);
+		}
 		ethqos->serdes_speed = SPEED_1000;
 		stmmac_pcs_ctrl_ane(priv, priv->ioaddr, 1, 0, 0);
 		break;
@@ -754,6 +828,9 @@ static int ethqos_configure_usxgmii(struct qcom_ethqos *ethqos)
 {
 	ethqos_set_func_clk_en(ethqos);
 
+	if (ethqos->use_domains)
+		qcom_ethqos_serdes_set_level(ethqos);
+
 	rgmii_updatel(ethqos, RGMII_BYPASS_EN, RGMII_BYPASS_EN, RGMII_IO_MACRO_BYPASS);
 	rgmii_updatel(ethqos, RGMII_CONFIG2_MODE_EN_VIA_GMII, 0, RGMII_IO_MACRO_CONFIG2);
 	rgmii_updatel(ethqos, SGMII_PHY_CNTRL0_2P5G_1G_CLK_SEL, BIT(5),
@@ -870,6 +947,98 @@ static void qcom_ethqos_serdes_powerdown(struct net_device *ndev, void *priv)
 	phy_exit(ethqos->serdes_phy);
 }
 
+static int qcom_ethqos_serdes_up(struct net_device *ndev, void *priv)
+{
+	struct qcom_ethqos *ethqos = priv;
+	int ret = 0;
+
+	ret = qcom_ethqos_domain_on(ethqos, PERF_SERDES);
+	if (ret < 0)
+		return ret;
+
+	ret = qcom_ethqos_serdes_set_level(ethqos);
+
+	return ret;
+}
+
+static void qcom_ethqos_serdes_down(struct net_device *ndev, void *priv)
+{
+	struct qcom_ethqos *ethqos = priv;
+
+	qcom_ethqos_domain_off(ethqos, PERF_SERDES);
+}
+
+/* callback for stmmac runtime suspend/resume functions */
+static int qcom_ethqos_domain_transition_d0d1(void *priv, bool high)
+{
+	struct qcom_ethqos *ethqos = priv;
+	int ret = 0;
+
+	if (high) {
+		ret = qcom_ethqos_domain_on(ethqos, POWER_MDIO);
+		if (ret < 0) {
+			dev_err(&ethqos->pdev->dev, "Transition from d1 to d0 failed\n");
+			return ret;
+		}
+		ethqos_set_func_clk_en(ethqos);
+	} else {
+		qcom_ethqos_domain_off(ethqos, POWER_MDIO);
+	}
+
+	return ret;
+}
+
+static int qcom_ethqos_domain_transition_d0d3(void *priv, bool high)
+{
+	struct qcom_ethqos *ethqos = priv;
+	int ret = 0;
+
+	if (high) {
+		ret = qcom_ethqos_domain_on(ethqos, POWER_CORE);
+		if (ret < 0) {
+			dev_err(&ethqos->pdev->dev, "CORE Transition from d3 to d0 failed\n");
+			return ret;
+		}
+
+		ret = qcom_ethqos_domain_on(ethqos, POWER_MDIO);
+		if (ret < 0) {
+			dev_err(&ethqos->pdev->dev, "MDIO Transition from d3 to d0 failed\n");
+			qcom_ethqos_domain_off(ethqos, POWER_CORE);
+			return ret;
+		}
+
+	} else {
+		qcom_ethqos_domain_off(ethqos, POWER_MDIO);
+		qcom_ethqos_domain_off(ethqos, POWER_CORE);
+	}
+
+	return ret;
+}
+
+static void qcom_ethqos_exit(struct platform_device *pdev, void *prv)
+{
+	struct net_device *ndev = platform_get_drvdata(pdev);
+	struct stmmac_priv *priv = netdev_priv(ndev);
+	struct qcom_ethqos *ethqos = prv;
+
+	/* pm_runtime_disable is called on driver remove.
+	 * And not called on Suspend/resume.
+	 */
+	if (!pm_runtime_enabled(priv->device)) {
+		pm_runtime_enable(priv->device);
+		qcom_ethqos_domain_transition_d0d3(prv, false);
+		pm_runtime_disable(priv->device);
+		dev_pm_domain_detach_list(ethqos->pd_list);
+	} else {
+		qcom_ethqos_domain_transition_d0d3(prv, false);
+	}
+}
+
+static int qcom_ethqos_init(struct platform_device *pdev, void *prv)
+{
+	return qcom_ethqos_domain_transition_d0d3(prv, true);
+}
+
 static int ethqos_clks_config(void *priv, bool enabled)
 {
 	struct qcom_ethqos *ethqos = priv;
@@ -1006,6 +1175,28 @@ static int qcom_ethqos_probe(struct platform_device *pdev)
 	}
 
 	ethqos->pdev = pdev;
+
+	if (of_device_is_compatible(np, "qcom,sa8255p-ethqos") ||
+		of_device_is_compatible(np, "qcom,sa8620p-ethqos") ||
+		of_device_is_compatible(np, "qcom,sa8797p-ethqos")) {
+		ret = qcom_ethqos_domain_attach(ethqos);
+		if (ret < 0) {
+			dev_err(dev, "Failed to attach domains.\n");
+			return ret;
+		}
+
+		dev_info(dev, "domains loaded successfully.\n");
+		ethqos->use_domains = true;
+		plat_dat->clks_config = qcom_ethqos_domain_transition_d0d1;
+		plat_dat->serdes_powerup = qcom_ethqos_serdes_up;
+		plat_dat->serdes_powerdown = qcom_ethqos_serdes_down;
+		plat_dat->exit = qcom_ethqos_exit;
+		plat_dat->init = qcom_ethqos_init;
+
+		qcom_ethqos_domain_transition_d0d3(ethqos, true);
+		qcom_ethqos_domain_on(ethqos, PERF_SERDES);
+	}
+
 	ethqos->rgmii_base = devm_platform_ioremap_resource_byname(pdev, "rgmii");
 	if (IS_ERR(ethqos->rgmii_base))
 		return dev_err_probe(dev, PTR_ERR(ethqos->rgmii_base),
@@ -1020,11 +1211,12 @@ static int qcom_ethqos_probe(struct platform_device *pdev)
 	ethqos->has_emac_ge_3 = data->has_emac_ge_3;
 	ethqos->needs_sgmii_loopback = data->needs_sgmii_loopback;
 
-	ethqos->link_clk = devm_clk_get(dev, data->link_clk_name ?: "rgmii");
-	if (IS_ERR(ethqos->link_clk))
-		return dev_err_probe(dev, PTR_ERR(ethqos->link_clk),
-				     "Failed to get link_clk\n");
-
+	if (!ethqos->use_domains) {
+		ethqos->link_clk = devm_clk_get(dev, data->link_clk_name ?: "rgmii");
+		if (IS_ERR(ethqos->link_clk))
+			return dev_err_probe(dev, PTR_ERR(ethqos->link_clk),
+						 "Failed to get link_clk\n");
+	}
 	ret = ethqos_clks_config(ethqos, true);
 	if (ret)
 		return ret;
@@ -1046,7 +1238,10 @@ static int qcom_ethqos_probe(struct platform_device *pdev)
 	plat_dat->bsp_priv = ethqos;
 	plat_dat->fix_mac_speed = ethqos_fix_mac_speed;
 	plat_dat->dump_debug_regs = rgmii_dump;
-	plat_dat->ptp_clk_freq_config = ethqos_ptp_clk_freq_config;
+	if (ethqos->use_domains)
+		plat_dat->clk_ptp_rate = data->ptp_clk_rate;
+	else
+		plat_dat->ptp_clk_freq_config = ethqos_ptp_clk_freq_config;
 	plat_dat->has_gmac4 = 1;
 	if (ethqos->has_emac_ge_3)
 		plat_dat->dwmac4_addrs = &data->dwmac4_addrs;
@@ -1089,6 +1284,7 @@ static int qcom_ethqos_probe(struct platform_device *pdev)
 static const struct of_device_id qcom_ethqos_match[] = {
 	{ .compatible = "qcom,qcs404-ethqos", .data = &emac_v2_3_0_data},
 	{ .compatible = "qcom,qcs8300-ethqos", .data = &emac_v4_0_0_data},
+	{ .compatible = "qcom,sa8255p-ethqos", .data = &emac_v4_0_0_data},
 	{ .compatible = "qcom,sa8775p-ethqos", .data = &emac_v4_0_0_data},
 	{ .compatible = "qcom,sc8280xp-ethqos", .data = &emac_v3_0_0_data},
 	{ .compatible = "qcom,sm8150-ethqos", .data = &emac_v2_1_0_data},
-- 
2.34.1

